src/hooks/unstable_useSlots.ts

```diff
@@ -0,0 +1,65 @@
+import React from 'react'
+
+type SlotConfig = Record<string, React.ComponentType>
+
+type SlotElements<Type extends SlotConfig> = {
+  [Property in keyof Type]: React.ReactElement
+}
+
+/**
+ * Extract components from `children` so we can render them in different places,
+ * allowing us to implement components with SSR-compatible slot APIs.
+ * Note: We can only extract direct children, not nested ones.
+ */
+// eslint-disable-next-line camelcase
+export function unstable_useSlots<T extends SlotConfig>(
+  children: React.ReactNode,
+  config: T,
+): [Partial<SlotElements<T>>, React.ReactNode[]] {
+  // Object mapping slot names to their elements
+  const slots: Partial<SlotElements<T>> = mapValues(config, () => undefined)
+
+  // Array of elements that are not slots
+  const rest: React.ReactNode[] = []
+
+  // eslint-disable-next-line github/array-foreach
+  React.Children.forEach(children, child => {
+    if (!React.isValidElement(child)) {
+      rest.push(child)
+      return
+    }
+
+    // Check if the child is a slot
+    const slotKey = getKeyByValue(config, child.type)
+
+    if (slotKey && slots[slotKey]) {
+      // If slot is already filled, ignore duplicates
```

Comment:
Would it be helpful here to emit a warning about a duplicate slot or would that just add noise to the console?

Reply:
Great idea! Done ✅